### PR TITLE
Adding a configuration flag for setting the allocation policy limit.

### PR DESCRIPTION
**Pull request by VSadov**
The flag is useful for testing purposes and for setting the policy limit in non-JSRT scenarios.
- Fixed some missing `ReportExternalMemoryFree` calls.
Not having those was tripping consistency asserts in memory policy manager in debug and could potentially result in false OOMs in release.

Fixes:https://github.com/Microsoft/ChakraCore/issues/5031

### CODE REVIEW

**Comment by VSadov** [comment]
@Cellule  - Michael, could you take a look at this change? In particular at the `ProjectionArrayBuffer` changes. 
It looks like we were not always reporting freeing of buffers consistently and it was resulting in asserts once we started using the allocation policy in more cases.

Thanks!!

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 1]
```diff
@@ -1024,9 +1026,35 @@ namespace Js
...
+    void ProjectionArrayBuffer::Finalize(bool isShutdown)
     {
+        if (isExternal)
+        {
+            return;
+        }
+
         CoTaskMemFree(buffer);
+
+        // In debugger scenario, ScriptAuthor can create scriptContext and delete scriptContext
```
Comment (curtisman) [inline lib/Runtime/Library/ArrayBuffer.cpp]: ScriptAuthor doesn't exist any more.
And it's always fine to get it from the library, especially in the Finalizer, as nothing is swept yet.

So you probably don't need this comment.

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 2]
```diff
@@ -1024,9 +1026,35 @@ namespace Js
         return RecyclerNewFinalized(recycler, ProjectionArrayBuffer, buffer, length, type);
     }
 
-    void ProjectionArrayBuffer::Dispose(bool isShutdown)
+    void ProjectionArrayBuffer::Finalize(bool isShutdown)
     {
+        if (isExternal)
```
Comment (curtisman) [inline lib/Runtime/Library/ArrayBuffer.cpp]: Why don't we want to do CoTaskMemFree if it is external?

Comment (curtisman) [reply lib/Runtime/Library/ArrayBuffer.cpp]: Before your change, we would have CoTaskMemFree that case too.

---
In reply to: [186237676](https://github.com/Microsoft/ChakraCore/pull/5099#discussion_r186237676) [](ancestors = 186237676)

File: lib/Runtime/Library/ArrayBuffer.cpp (hunk 1 above)
Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: I thought ScripAuthor is the user :-), did not know it was a thing. 
I just copied this piece from another finalizer.

I will modify comments in both places

File: lib/Runtime/Library/ArrayBuffer.cpp (hunk 2 above)
Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: Right, I was not sure about the ownership here. It did look as something that might be owned externally.
As we discussed we will be treating external buffer as completely owned. - will always free in finalize and charge/uncharge to the quota regardless who allocated.

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 3]
```diff
@@ -707,7 +713,12 @@ namespace Js
         {
             freeFunction(this->buffer);
             this->buffer = nullptr;
-            this->recycler->ReportExternalMemoryFree(this->bufferLength);
+
+            // for projection array buffers we have already reported freeing on detach
+            if (this->allocationType != ArrayBufferAllocationType::CoTask)
```
Comment (VSadov) [inline lib/Runtime/Library/ArrayBuffer.cpp]: @Cellule - I had to do a small change here. Comparing template functions would not build in Core, so I changed this to compare `allocationType`

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 4]
```diff
@@ -218,6 +218,12 @@ namespace Js
     {
         Assert(!this->isDetached);
 
+        if (this->IsProjectionArrayBuffer())
+        {
+            Recycler* recycler = GetType()->GetLibrary()->GetRecycler();
```
Comment (curtisman) [inline lib/Runtime/Library/ArrayBuffer.cpp]: why only projection array buffer?

Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: Right. It is logical to report free on any Detach, but that was causing failures and we suspected a separate bug, possibly in a test/library.
Now it looks like we just have the same issue in `JavascriptArrayBuffer` as the one that was fixed for projection buffers - it does not report taking posession of external buffers, but does report free when finalized. If that is fixed, we might not need to specialcase projection buffers.

Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: Well, it looks like `WebAssemblyArrayBuffer` also needs fixing. When it grows, it charges for the  _diff_ in sizes, which noone reports freeing later.

File: lib/Runtime/Library/ArrayBuffer.cpp (hunk 3 above)
Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: no longer applicable

---
In reply to: [186249460](https://github.com/Microsoft/ChakraCore/pull/5099#discussion_r186249460) [](ancestors = 186249460)

**Comment by VSadov** [comment]
I think the whole ownership of the externally allocated memory and allocation policy needs some clean-up. It is a bit ad-hoc right now and it is easy to make mistakes.

File: lib/Runtime/Library/ArrayBuffer.h [hunk 5]
```diff
@@ -325,7 +326,8 @@ namespace Js
         // take over ownership. a CoTaskMemAlloc'ed buffer passed in via projection.
         static ProjectionArrayBuffer* Create(byte* buffer, DECLSPEC_GUARD_OVERFLOW uint32 length, DynamicType * type);
         virtual void Dispose(bool isShutdown) override;
-        virtual void Finalize(bool isShutdown) override {};
+        virtual void Finalize(bool isShutdown) override;
+        virtual bool IsProjectionArrayBuffer() override { return true; }
```
Comment (akroshg) [inline lib/Runtime/Library/ArrayBuffer.h]: >IsProjectionArrayBuffer [](start = 21, length = 23)

Trying to find out where is this is used ?

File: lib/Common/ConfigFlagsList.h [hunk 6]
```diff
@@ -1430,6 +1431,7 @@ FLAGNR(Boolean, RecyclerVerifyMark    , "verify concurrent gc", false)
 #endif
 FLAGR (Number,  LowMemoryCap          , "Memory cap indicating a low-memory process", DEFAULT_CONFIG_LowMemoryCap)
 FLAGNR(Number,  NewPagesCapDuringBGSweeping, "New pages count allowed to be allocated during background sweeping", DEFAULT_CONFIG_NewPagesCapDuringBGSweeping)
+FLAGR (Number,  AllocPolicyLimit      , "Memory allocation policy limit in MB (default: -1).", DEFAULT_CONFIG_AllocationPolicyLimit)
```
Comment (akroshg) [inline lib/Common/ConfigFlagsList.h]: >default: -1 [](start = 79, length = 11)

nit: should be better to word this as 
(default: -1, which means no allocation policy limit)
?

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 7]
```diff
@@ -696,6 +701,13 @@ namespace Js
         Recycler* recycler = type->GetScriptContext()->GetRecycler();
         JavascriptArrayBuffer* result = RecyclerNewFinalized(recycler, JavascriptArrayBuffer, buffer, length, type);
         Assert(result);
+
+        // we take the ownership of the buffer and will have to free it so charge it to our quota.
+        if (!recycler->RequestExternalMemoryAllocation(length))
```
Comment (akroshg) [inline lib/Runtime/Library/ArrayBuffer.cpp]: why not do this in ArrayBuffer's constructor?

File: lib/Runtime/Library/ArrayBuffer.h (hunk 5 above)
Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.h]: This is no longer needed. It was use in the iteration where we special cased projection buffers.

---
In reply to: [186509650](https://github.com/Microsoft/ChakraCore/pull/5099#discussion_r186509650) [](ancestors = 186509650)

File: lib/Runtime/Library/ArrayBuffer.cpp (hunk 7 above)
Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: I was not sure about getting access to recycler while `this` is still not constructed. And was not comfortable with throwing exceptions from a constructor too.

I plan to revisit the whole deal with allocation of external memory and allocation policy. This code may change anyways.
For now we just want to fix what we have.

---
In reply to: [186512931](https://github.com/Microsoft/ChakraCore/pull/5099#discussion_r186512931) [](ancestors = 186512931)

Comment (VSadov) [reply lib/Runtime/Library/ArrayBuffer.cpp]: Added a tracking bug to get to this later - [https://github.com/Microsoft/ChakraCore/issues/5114](https://github.com/Microsoft/ChakraCore/issues/5114)


---
In reply to: [186617696](https://github.com/Microsoft/ChakraCore/pull/5099#discussion_r186617696) [](ancestors = 186617696,186512931)

**Pull request merged**